### PR TITLE
feat: Execute macro actions, for the conversation

### LINKED ISSUE #5013: Add an API to execute macros

### PR DESCRIPTION
**Pull request by tejaswinichile**
# Pull Request Template

## Description

Execute macros actions for the conversation.

Fixes #5013 

## Type of change

Please delete options that are not relevant.

- [ ] New feature (non-breaking change which adds functionality)

## How Has This Been Tested?

Tested on local and with test cases.

## Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented on my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

https://user-images.githubusercontent.com/10827585/179926806-d3e6be1f-ab4e-4202-9ed3-745b2c103707.mov

### CODE REVIEW

**Comment by pranavrajs** [comment]
@sojan-official Yes, this is correct. The user context should be used. Unlike automation, the action is performed by the agent, so every action should consider the user context (e.g., A message should be sent in the name of the agent, or labels should have the agent name, etc.). Also, we need support for multiple conversation ids in the API. 

@tejaswinichile Noticed that the add_label action didn't work in the screencast.

**Comment by tejaswinichile** [comment]
> @tejaswinichile Noticed that the add_label action didn't work in the screencast.

@pranavrajs Yeah because I didn't have those labels added to my account, it works, let me add the updated video.

> Also, for the macros endpoint, we will need to accept an array of conversation ids if we are thinking about having it in bulk actions.

@pranavrajs @sojan-official Should I update the PR for the bulk actions?

**Comment by tejaswinichile** [comment]
https://user-images.githubusercontent.com/10827585/180042477-776c159c-3c94-4116-b7f8-e237e3b61f04.mov

Hey, @pranavrajs @sojan-official Please have a look at this updated video.

**Comment by pranavrajs** [comment]
> @pranavrajs @sojan-official Should I update the PR for the bulk actions?

I guess you can update the conversation_id parameter in the API to use `conversation_ids`.

File: app/controllers/api/v1/accounts/macros_controller.rb [hunk 1]
```diff
@@ -34,6 +34,12 @@ def update
     end
   end
 
+  def execute
+    ::MacrosExecutionJob.perform_later(@macro, conversation_ids: params[:conversation_ids], user: current_user)
```
Comment (sojan-official) [inline app/controllers/api/v1/accounts/macros_controller.rb:38]: let's use `Current.user` instead. We are trying to remove the usage `current_user` in favor of this.

File: app/services/macros/execution_service.rb [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+
+  def add_label(labels)
+    return if labels.empty?
+
+    @conversation.add_labels(labels)
+  end
+
+  def send_email_to_team(_params); end
+
+  def agent_belongs_to_account?(agent_ids)
```
Comment (sojan-official) [inline app/services/macros/execution_service.rb:75]: some of these methods seem repetitive to what we have in automation service 
https://github.com/chatwoot/chatwoot/blob/c4b2005425f7a02003fb03d80bc45f445a50114b/app/services/automation_rules/action_service.rb#L101

can we extract them to a common base class or a helper and avoid the repetition?

**Review by sojan-official — COMMENTED** [review]
Added a couple of comments. Looks good overall.

File: app/services/macros/execution_service.rb (hunk 2 above)
Comment (tejaswinichile) [reply app/services/macros/execution_service.rb:75]: Yes, we can have a common action service may be. The issue was just with the activity execution, let me check.

**Comment by tejaswinichile** [comment]
@sojan-official This has been fixed please have a look.

**Review by sojan-official — APPROVED** [review]
LGTM

**Pull request merged**